lib/css-converter.js

```diff
@@ -0,0 +1,311 @@
+import { CssSelectorParser } from 'css-selector-parser';
+import log from './logger';
+
+const CssConverter = {};
+
+const parser = new CssSelectorParser();
+parser.registerSelectorPseudos('has');
+parser.registerNestingOperators('>', '+', '~');
+parser.registerAttrEqualityMods('^', '$', '*', '~');
+parser.enableSubstitutes();
+
+const BOOLEAN_ATTRS = [
+  'checkable', 'checked', 'clickable', 'enabled', 'focusable',
+  'focused', 'long-clickable', 'scrollable', 'selected',
+];
+
+const NUMERIC_ATTRS = [
+  'index', 'instance',
+];
+
+const STR_ATTRS = [
+  'description', 'resource-id', 'text', 'class-name', 'package-name'
+];
+
+const ALL_ATTRS = [
+  ...BOOLEAN_ATTRS,
+  ...NUMERIC_ATTRS,
+  ...STR_ATTRS,
+];
+
+const attributeAliases = [
+  ['resource-id', ['id']],
+  ['description', [
+    'content-description', 'content-desc',
+    'desc', 'accessibility-id',
+  ]],
+  ['index', ['nth-child']],
+];
+
+const RESOURCE_ID = 'resource-id';
+
+/**
+ * Convert hyphen separated word to snake case
+ *
+ * @param {string} str
+ * @returns {string} The hyphen separated word translated to snake case
+ */
+function toSnakeCase (str) {
+  const tokens = str.split('-').map((str) => str.length > 0 ? str.charAt(0).toUpperCase() + str.slice(1).toLowerCase() : '');
+  let out = tokens.join('');
+  return out.charAt(0).toLowerCase() + out.slice(1);
+}
+
+/**
+ * Get the boolean from a CSS object. If empty, return true. If not true/false/empty, throw exception
+ *
+ * @param {Object} css CSS object
+ * @returns {string} Either 'true' or 'false'. If value is empty, return 'true'
+ */
+function assertGetBool (css) {
+  const val = css.value?.toLowerCase() || 'true'; // an omitted boolean attribute means 'true' (e.g.: input[checked] means checked is true)
+  if (['true', 'false'].includes(val)) {
+    return val;
+  }
+  log.errorAndThrow(`Could not parse '${css.name}=${css.value}'. '${css.name}' must be true, false or empty`);
+}
+
+/**
+ * Get the canonical form of a CSS attribute name
+ *
+ * Converts to lowercase and if an attribute name is an alias for something else, return
+ * what it is an alias for
+ *
+ * @param {Object} css CSS object
```

Comment:
I would rather prefer more specific definition of the CSS object with properties. Something like https://github.com/appium/appium-uiautomator2-driver/blob/f304d2c166c94bb721a884d2c96b5fda80d98a0a/lib/commands/alert.js#L8